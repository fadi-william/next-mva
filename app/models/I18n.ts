// This is a mst store to manage i18n in the application.
import { applySnapshot, types } from "mobx-state-tree";
import { addLocaleData } from "react-intl";

// Import the i18n local storage module.
import { getCurrentUserLocale, setCurrentUserLocale } from "../storage/i18n";

// Currently, we support english and french.
import * as enLocaleData from "react-intl/locale-data/en";
import * as frLocaleData from "react-intl/locale-data/fr";

// Import the i18n json files.
import * as enTranslatedData from "../../lang/en.json";
import * as frTranslatedData from "../../lang/fr.json";

export type TI18n = typeof I18n.Type;

// The module's global variables.
let store: TI18n = null;
let currentLocale = null;

export const I18n = types
  .model({
    locale: types.string,
    messages: types.optional(types.frozen, {})
  })
  .actions(self => ({
    loadLocale(lang: string) {
      self.locale = lang;
      setCurrentUserLocale(lang);
      if (lang === "en") {
        addLocaleData(enLocaleData);
        self.messages = enTranslatedData;
      } else {
        addLocaleData(frLocaleData);
        self.messages = frTranslatedData;
      }
    }
  }))
  .actions(self => ({
    afterCreate() {
      self.loadLocale(self.locale);
    }
  }));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(locale?, snapshot = null) {
  if (!locale) {
    locale = getCurrentUserLocale();
    currentLocale = locale;
  }
  if (store === null || currentLocale !== locale) {
    store = I18n.create({ locale });
  }
  if (snapshot) {
    applySnapshot(store, snapshot);
  }
  return store;
}
