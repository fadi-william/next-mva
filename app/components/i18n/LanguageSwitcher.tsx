// The imported libs.
import { observer } from "mobx-react";
import React from "react";

// Import the mst store.
import { getStore } from "../../models/I18n";

@observer
class LanguageSwitcher extends React.Component {
  public render() {
    const store = getStore();

    return (
      <select
        defaultValue={store.locale}
        onChange={event => {
          store.loadLocale(event.target.value);
        }}
      >
        <option value="fr">French</option>
        <option value="en">English</option>
      </select>
    );
  }
}

export default LanguageSwitcher;
