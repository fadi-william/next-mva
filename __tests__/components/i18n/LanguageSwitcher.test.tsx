import * as React from "react";
import * as renderer from "react-test-renderer";

// Import Enzyme.
import Enzyme from "../../../utils/enzyme";

import LanguageSwitcher from "../../../app/components/i18n/LanguageSwitcher";

// The mst store.
import { getStore } from "../../../app/models/I18n";

// Mock the i18n local storage module.
jest.mock("../../../app/storage/i18n");

it("test language switcher component - snapshot test", () => {
  const tree = renderer.create(<LanguageSwitcher />).toJSON();

  expect(tree).toMatchSnapshot();
});

it("test language switcher component's select options", () => {
  const wrapper = Enzyme.mount(<LanguageSwitcher />);
  expect(wrapper.find("select>option").length).toBe(2);
});

it("test language switcher component in action", () => {
  const wrapper = Enzyme.mount(<LanguageSwitcher />);
  expect(wrapper.find("select>option").length).toBe(2);

  expect(wrapper.find("select").prop("defaultValue")).toBe("en");
  wrapper.find("select").simulate("change", { target: { value: "fr" } });
  expect(wrapper.find("select").prop("defaultValue")).toBe("fr");

  const store = getStore();
  expect(store.locale).toBe("fr");
});
