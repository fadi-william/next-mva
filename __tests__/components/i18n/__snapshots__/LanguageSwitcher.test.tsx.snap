// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test language switcher component - snapshot test 1`] = `
<select
  defaultValue="en"
  onChange={[Function]}
>
  <option
    value="fr"
  >
    French
  </option>
  <option
    value="en"
  >
    English
  </option>
</select>
`;
